### PR TITLE
Added render and rendercomplete events to renderPanel method

### PR DESCRIPTION
**Pull request by umbe1987**
Should add what's discussed in #209.
Please feel free to take your time and see if it fits your project.
Also....sorry for taking so much time!!!

### CODE REVIEW

**Comment by umbe1987** [comment]
Hi @walkermatt 
Did you by any chance have the time to look into this PR or are willing to review it?
If you need further details or anything else, I am available 😉

**Comment by walkermatt** [comment]
@umbe1987, sorry it's taken so long to look at this.

I feel as though the component itself should dispatch the events, not the panel element. I'd prefer it if we should use the OpenLayers event system via [`dispatchEvent`](https://openlayers.org/en/latest/apidoc/module-ol_events_Target-Target.html#dispatchEvent) (which is available as `LayerSwitcher` inherits from `Control`). This would mean dispatching the event from the `renderPanel` instance method instead of the static.

Something like this 

```js
    /**
     * Re-draw the layer panel to represent the current state of the layers.
     */
    renderPanel() {
        this.dispatchEvent({ type: 'render' });
        LayerSwitcher.renderPanel(this.getMap(), this.panel, {
            groupSelectStyle: this.groupSelectStyle
        });
        this.dispatchEvent({ type: 'rendercomplete' });
    }
```

To listen to the event you'd do the following in your application:

```js
var layerSwitcher = new LayerSwitcher();

layerSwitcher.on('rendercomplete', function (evt) {
    console.log('LayerSwitcher has rendered...');
});

map.addControl(layerSwitcher);
```
If you're calling the static `renderPanel` yourself (as I think you might be in one of your applications) then you probably don't need an event as you can apply whatever logic is required before/ after calling `LayerSwitcher.renderPanel`.

**Comment by umbe1987** [comment]
Hi @walkermatt 
You're totally right that it should be implemented in the component itself.

> If you're calling the static renderPanel yourself (as I think you might be in one of your applications) then you probably don't need an event as you can apply whatever logic is required before/ after calling LayerSwitcher.renderPanel.

About this, I am not 100% sure it's gonna work as you said.
Let me explain: the reason I needed to add the event is that `renerPanel` is calling itself recursively (I guess it has to do with the `groupSelectStyle` implementation).

https://github.com/walkermatt/ol-layerswitcher/blob/d7014be166270afeba46200d77e51d2753e8710a/src/ol-layerswitcher.js#L155-L159

The point is: in my application (but I bet it's not the only case), I need to do something (display legend, zoom, attribute table icons), whenever the panel is redrawn. As far as I can tell, this is (also) happening whenever I check/uncheck a group layer. I cannot take actions when renderPanel calls itself recursively, hence my need of an event.

Hope this makes sense to you, I'll be happy to rework on this PR as you suggested (which anyway is way better than what I did ;)).

**Comment by umbe1987** [comment]
Quoting and correcting myself:
> Let me explain: the reason I needed to add the event is that renerPanel is calling itself recursively.

This is NOT the point. The point is that whenever an input checkbox changes, the `render` function defined within the `renderPanel` is called and this redraws the panel.

So, these lines:

https://github.com/walkermatt/ol-layerswitcher/blob/d7014be166270afeba46200d77e51d2753e8710a/src/ol-layerswitcher.js#L306-L315

trigger these lines:

https://github.com/walkermatt/ol-layerswitcher/blob/d7014be166270afeba46200d77e51d2753e8710a/src/ol-layerswitcher.js#L155-L159

So, I guess my point of having a `'render'` event might still be valid.

**Comment by umbe1987** [comment]
@walkermatt 
With bdab686383ffa0d998a7e91b3034072f855caa68 and 582f9c1457e48ce912758a66368fdf5474127ada I tried to implement a logic that should work both for the static and the Class `renderPane()l` methods.
Hope it is acceptable for merge.

The logic is:

- there's a new class method called `panelEvent()` that is used to dispatch both `'render'` and `'rendercomplete'` events
- it returns `true` when it's called with `'render'` or `'rendercomplete'` argument.
-  **static** `renderPanel()` method has a new optional `panel_event` param that defaults to `false`.
- when  **static** `renderPanel()` is called, `panel_event` is default (`false`) and so `'render'` and `'rendercomplete'` events are dispatched at the beginning and at the end of the function.
- when **dynamic** `renderPanel()` method is called, a call to `panelEvent("render")` happens before the call to the internal static `renederPanel()` and the param `panel_event` is passed as `true`, making the static `renderPanel()` not re-dispatching the event. A call to `panelEvent("rendercomplete")` happens at the end so both events are dispatched.

As a result, one can:

- do `layerSwitcher.on('render*')` to listen to the new events when using the dynamic `renderPanel()`.
- do `document.addEventListener("render*", function() {//DO SOMETHING});` when using the static `renderPanel()`.

**Comment by walkermatt** [comment]
I'm happy dispatching events via the instance and the panel element. Couple of thoughts:

* I think the element event should be dispatched by the panel element instead of the document
* I don't think we need to worry about avoiding dispatching the element events.

**Comment by umbe1987** [comment]
> I think the element event should be dispatched by the panel element instead of the document

Totally agree :thumbsup:

> I don't think we need to worry about avoiding dispatching the element events.

I hope to have understood correctly.

I know it's a bit of new rows in the code to do that, but in some cases if many things have to occur each time the panel gets rendered, IMHO it is better to dispatch the event only when it's necessary.

But I'm fine with removing the switch case in the `panelEvent()` if that's what you meant.

**Comment by umbe1987** [comment]
@walkermatt is it fine for merge after the last commit?

File: src/ol-layerswitcher.js [hunk 1]
```diff
@@ -109,21 +109,48 @@ export default class LayerSwitcher extends Control {
...
+    }
+
     /**
     * Re-draw the layer panel to represent the current state of the layers.
     */
     renderPanel() {
-        LayerSwitcher.renderPanel(this.getMap(), this.panel, {
-            groupSelectStyle: this.groupSelectStyle
-        });
+        var panel_event;
```
Comment (walkermatt) [inline src/ol-layerswitcher.js]: Not sure we need the `panel_event` argument or variable. It wouldn't hurt to always dispatch.

File: src/ol-layerswitcher.js [hunk 2]
```diff
@@ -109,21 +109,48 @@ export default class LayerSwitcher extends Control {
...
+        panel_event = this.panelEvent("render");
```
Comment (walkermatt) [inline src/ol-layerswitcher.js]: I think we can also drop the `panelEvent` method, simply call `this.dispatchEvent({ type: 'render' })` here?

File: src/ol-layerswitcher.js [hunk 3]
```diff
@@ -109,21 +109,48 @@ export default class LayerSwitcher extends Control {
...
+        LayerSwitcher.renderPanel(this.getMap(), this.panel, panel_event, { groupSelectStyle: this.groupSelectStyle });
+        this.panelEvent("rendercomplete");
```
Comment (walkermatt) [inline src/ol-layerswitcher.js]: Same here `this.dispatchEvent({ type: 'rendercomplete' });` feels sufficient

File: src/ol-layerswitcher.js (hunk 1 above)
Comment (umbe1987) [reply src/ol-layerswitcher.js]: Agree, removed!

File: src/ol-layerswitcher.js (hunk 3 above)
Comment (umbe1987) [reply src/ol-layerswitcher.js]: Yep, I thought all the panelEvent() method was necessary not to dispatch too many events, but indeed it is working fine without. Removed.

File: src/ol-layerswitcher.js (hunk 2 above)
Comment (umbe1987) [reply src/ol-layerswitcher.js]: Yep, I thought all the panelEvent() method was necessary not to dispatch too many events, but indeed it is working fine without. Removed.

File: src/ol-layerswitcher.js (hunk 3 above)
Comment (umbe1987) [reply src/ol-layerswitcher.js]: Agree

**Comment by umbe1987** [comment]
Last commit should do the same as before in a much simpler manner. Thanks for your review!

**Comment by walkermatt** [comment]
@umbe1987 merged in 32735005cb5eaa3c20310856dc346fb0261f7c4c :tada: 

Could you follow up with an example that uses the events to enhance the layer switcher in the way that you are doing in your app. Maybe add a couple of opacity sliders or similar?

Can you also please add some docs.

**Comment by umbe1987** [comment]
Great that this has been merged!

`Could you follow up with an example that uses the events to enhance the layer switcher in the way that you are doing in your app. Maybe add a couple of opacity sliders or similar?`

Sure, as soon as I'll find some time to do it, I'll be happy to include an example with opacity sliders.

**Comment by umbe1987** [comment]
@walkermatt 

Ehi, I am working on implementing the example including the opacity sliders (you can see it [here](http://raw.githack.com/umbe1987/ol-layerswitcher/renderevent_example/examples/render_event.html)).

When the panel opens everything is fine.

For some reason, I cannot seem to have it working when checking/unchecking the layers when the panel is opened.

I guess it still has to do with the panel re-rendering when doing this, but this was one of the reason I added the render events, so no clue what's going on.

Also, don't know why, but cloning the repo wasn't sufficient to have the render and rendercomplete inside the code: I had to copy and paste both the js in src and dist folder to have them. Strange...

Anyway, hope to figure out what's going on :crossed_fingers:

Cheers!

**Comment by umbe1987** [comment]
More precisely, debugging the code, it seems the problem is in the call to render(lyr) in these lines:

https://github.com/walkermatt/ol-layerswitcher/blob/cdcb827cf503f0693d2e0d97f6b39ddcdeec38bb/src/ol-layerswitcher.js#L355

This is calling `LayerSwitcher.renderPanel(map, panel, options);` which should dispatch the render events another time, but this does not happen and this is the problem:

https://github.com/walkermatt/ol-layerswitcher/blob/cdcb827cf503f0693d2e0d97f6b39ddcdeec38bb/src/ol-layerswitcher.js#L163-L166

Placing a breakpoint at the call to render(lyr); and checking/unchecking one of the layer checkboxes, this can be clearly seen: the renderPanel method is called but the render events are not dispatched.

Hopefully I will be able to make an animation to show this clearly.

**Pull request closed without merging**